### PR TITLE
Simple Staking Payouts

### PR DESCRIPTION
**Pull request by shawntabrizi**
This is a simplified version of Flexible Payouts (https://github.com/paritytech/substrate/pull/5268).

Here we keep nearly everything about the existing staking pallet the same, reducing the total number of changes introduced by an alternative payout mechanism.

The changes introduced are:

* We track which specific eras have been rewarded between `current_era - history_depth` and `current_era`. This means that rewards can now be claimed out of order without issue.

* We added a new extrinsic `payout_stakers` which takes a validator and era as an input, and pays out the validator and all their nominators at once.

* We allow any signed origin to call `payout_stakers`. This means any account is able to pay the fees to trigger a payout for a validator and their nominators.

* We "deprecated" `payout_nominator` and `payout_validator`, see the migration story below.

As a result of this PR, everyone who would be paid currently still gets paid. The only difference is that nominators will not need to manually claim their rewards. Validators payouts will do that for them automatically.

- [x] Comprehensive tests written
- [x] Documentation Updates
- [x] Migration Code

EDIT: Migration Story

We will use a "two-update" migration story to implement this new payout system.

The issue is that our current payment mechanism allows for nominators and validators to be out of sync in terms of payments, whereas this payment mechanism will keep them in sync.

When transitioning from the old mechanism to this mechanism, we need to find a way to bring the nominators and validators in sync. One approach would be to make payments to everyone for every era up to the current era, but I felt that this was a heavy and possibly error prone operation.

Rather, I decided to keep the old extrinsics in the codebase for now, but place a check that the old extrinsics could only be called to make payouts for eras before this update. Then, when new eras come, and people will request payouts, they will have to use the new extrinsic.

At some point, the old extrinsics will not be callable anymore (the upgrade era will be older than history depth), then we can do a simple second upgrade to remove the old extrinsics and any old logic.

The one thing we do need to migrate is the `last_reward` `Option` to `claimed_rewards` `Vec`. We do this for all nominators and validators for now.

When we remove the old extrinsics, we can also go through all the nominators and clean up their `claimed_rewards` vec, since it won't be used in the new logic.

### CODE REVIEW

File: frame/staking/src/lib.rs [hunk 1]
```diff
@@ -453,7 +453,7 @@ pub struct StakingLedger<AccountId, Balance: HasCompact> {
 	/// of the stash (assuming it doesn't get slashed first).
 	pub unlocking: Vec<UnlockChunk<Balance>>,
 	/// The latest and highest era which the staker has claimed reward for.
-	pub last_reward: Option<EraIndex>,
+	pub claimed_rewards: Vec<EraIndex>,
```
Comment (jacogr) [inline frame/staking/src/lib.rs]: Probably just need a doc update here (it is now the range)

**Comment by shawntabrizi** [comment]
I did basic migration verification in the form of a test, but I struggled to set up a production test, and it was eating at some of my time.

Would appreciate any quick help here to verify the migration is sane for Kusama accounts.

File: frame/staking/src/lib.rs [hunk 2]
```diff
@@ -1281,7 +1289,7 @@ decl_module! {
 				total: value,
 				active: value,
 				unlocking: vec![],
-				last_reward: Self::current_era(),
+				claimed_rewards: vec![],
```
Comment (thiolliere) [inline frame/staking/src/lib.rs]: Can the following scenario happen ?
* A validator have produced some blocks,
* the validator claims some reward,
* the validator unbond himself
* the validator bond himself
* the validator claim its reward again as claimed_reward is empty.

anyway you can say that all era in `[last_payout_era, active_era]` are claimed

File: frame/staking/src/tests.rs [hunk 3]
```diff
@@ -828,7 +827,7 @@ fn reward_destination_works() {
 		<Module<Test>>::reward_by_ids(vec![(11, 1)]);
 
 		start_era(1);
-		mock::make_all_reward_payment(0);
```
Comment (thiolliere) [inline frame/staking/src/tests.rs]: actually I prefer having these functions explicitly imported from mock so it shows this is just util function but I'm ok.

File: frame/staking/src/lib.rs [hunk 4]
```diff
@@ -2013,6 +2107,108 @@ impl<T: Trait> Module<T> {
 		Ok(())
 	}
 
+	fn do_payout_stakers(
+		validator_stash: T::AccountId,
+		era: EraIndex,
+	) -> DispatchResult {
+		/* Validate input data */
+		let current_era = CurrentEra::get().unwrap_or(0);
```
Comment (thiolliere) [inline frame/staking/src/lib.rs]: If CurrentEra is None then this means there is no current era thus I think it should return Error::InvalidaEraToReward

File: frame/staking/src/lib.rs (hunk 2 above)
Comment (shawntabrizi) [reply frame/staking/src/lib.rs]: ah good catch, will update

File: frame/staking/src/tests.rs (hunk 3 above)
Comment (shawntabrizi) [reply frame/staking/src/tests.rs]: I added it back, you make a good point.

File: frame/staking/src/lib.rs (hunk 2 above)
Comment (shawntabrizi) [reply frame/staking/src/lib.rs]: fixed and added tests

File: frame/staking/src/lib.rs (hunk 4 above)
Comment (shawntabrizi) [reply frame/staking/src/lib.rs]: fixed

**Comment by shawntabrizi** [comment]
I have updated the original post with our two-step migration plan. This should be ready for final reviews and merge!

**Comment by gavofyork** [comment]
@shawntabrizi needs resolving.

**Comment by gregzaitsev** [comment]
> The only difference is that nominators will not need to manually claim their rewards. Validators payouts will do that for them automatically.

Is it somehow guaranteed that validators call payout_stakers or you just assume that they will call it because their rewards are much larger than average nominator rewards?

**Comment by thiolliere** [comment]
> Is it somehow guaranteed that validators call payout_stakers or you just assume that they will call it because their rewards are much larger than average nominator rewards?

there is currently no mecanism provided to make it automatic, someone must call it, it can be anybody. Though it is the only way for the validator to gets its reward, but indeed it is also the only way for a nominator to get the part associated.

In the future we could think that the transactions fees are given back to the caller by taking from the payout but for no nothing is planned as far as I know.

**Pull request merged**